src/auditwheel/_vendor/whichprovides/__init__.py

```diff
@@ -0,0 +1,359 @@
+# SPDX-License-Identifier: MIT
+
+"""
+Module which provides (heh) 'yum provides'
+functionality across many package managers.
+"""
+
+import dataclasses
+import pathlib
+import re
+import shutil
+import subprocess
+import sys
+import typing
+from urllib.parse import quote
+
+__all__ = ["ProvidedBy", "whichprovides"]
+__version__ = "0.4.0"
+
+_OS_RELEASE_LINES_RE = re.compile(r"^([A-Z_]+)=(?:\"([^\"]*)\"|(.*))$", re.MULTILINE)
+_APK_WHO_OWNS_RE = re.compile(r" is owned by ([^\s\-]+)-([^\s]+)$", re.MULTILINE)
+_DPKG_SEARCH_RE = re.compile(r"^([^:]+):")
+_DPKG_VERSION_RE = re.compile(r"^Version: ([^\s]+)", re.MULTILINE)
+_APT_FILE_SEARCH_RE = re.compile(r"^([^:]+): (.+)$", re.MULTILINE)
+
+
+@dataclasses.dataclass
+class ProvidedBy:
+    package_type: str
+    package_name: str
+    package_version: str
+    distro: typing.Union[str, None] = None
+
+    @property
+    def purl(self) -> str:
+        """The Package URL (PURL) of the providing package"""
+
+        def _quote_purl(value: str) -> str:
+            """
+            Quotes according to PURL rules which are different from
+            typical URL percent encoding.
+            """
+            return quote(value, safe="")
+
+        # PURL disallows many characters in the package type field.
+        if not re.match(r"^[a-zA-Z0-9\+\-\.]+$", self.package_type):
+            raise ValueError("Package type must be ASCII letters, numbers, +, -, and .")
+
+        parts = ["pkg:", self.package_type.lower(), "/"]
+        if self.distro:
+            parts.extend((_quote_purl(self.distro), "/"))
+        parts.extend(
+            (_quote_purl(self.package_name), "@", _quote_purl(self.package_version))
+        )
+        return "".join(parts)
+
+
+class PackageProvider:
+    # Order in which the provider should be resolved.
+    # Lower is attempted earlier than higher numbers.
+    _resolve_order: int = 0
+    _has_bin_cache: dict[str, typing.Union[str, bool]] = {}
+
+    @staticmethod
+    def os_release() -> dict[str, str]:
+        """Dumb method of finding os-release information."""
+        try:
+            with open("/etc/os-release") as f:
+                os_release = {}
+                for name, value_quoted, value_unquoted in _OS_RELEASE_LINES_RE.findall(
+                    f.read()
+                ):
+                    value = value_quoted if value_quoted else value_unquoted
+                    os_release[name] = value
+                return os_release
+        except OSError:
+            return {}
+
+    @staticmethod
+    def distro() -> typing.Optional[str]:
+        return PackageProvider.os_release().get("ID", None)
```

Comment:
should this be cached or maybe `_package_providers` should be cached and only return available providers (or add a cached method `_available_package_providers()` ?

Reply:
Created as a PR to whichprovides here: https://github.com/sethmlarson/whichprovides/pull/2

Reply:
given it's only run once as mentioned in the upstream PR, this is fine as-is